x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -653,6 +658,11 @@ protected LogicalPlan rule(Limit limit) {
                         }
                     }
                 }
+            } else if (limit.child() instanceof Join join) {
```

Comment:
This was actually a suggestion from Costin. I didn't start this way. I do think it'd be useful to do things like push down things to the `LocalRelation`, but this is something I've got to trust the QL experts in.

Reply:
As discussed, I think a good compromise would be to only implement the specialized `LeftJoin` for now - so the fact that it's a left join is encoded in the class rather than in an external config. That allows to implement methods on it which may not apply to general joins, and we can more easily treat `Lookup` similarly to `Enrich` in optimization rules.